server/src/main/java/io/envoyproxy/controlplane/server/limits/StreamLimiter.java

```diff
@@ -0,0 +1,85 @@
+package io.envoyproxy.controlplane.server.limits;
+
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StreamLimiter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StreamLimiter.class);
+
+  private final long maxStreams;
+
+  private final ServerInterceptor serverInterceptor;
+
+  private final AtomicLong openStreams = new AtomicLong();
+
+  /**
+   * StreamLimiter is responsible for configuring a {@link ServerInterceptor} instance,
+   * which will take care of limiting concurrently open streams.
+   * @param maxStreams the upper limit for concurrently open streams
+   */
+  public StreamLimiter(long maxStreams) {
+    this.maxStreams = maxStreams;
+    this.serverInterceptor = new ServerInterceptor() {
+      @Override
+      public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+          ServerCall<ReqT, RespT> call,
+          Metadata headers,
+          ServerCallHandler<ReqT, RespT> next
+      ) {
+        if (openStreams.get() >= maxStreams) {
+          call.close(Status.UNAVAILABLE, new Metadata());
+          return new ServerCall.Listener<ReqT>() {};
+        }
+
+        final Listener<ReqT> delegate = next.startCall(call, headers);
+
+        openStreams.incrementAndGet();
+
+        LOGGER.debug("stream opened [open/max: {} / {}]", openStreams.get(), maxStreams);
+
+        return new Listener<ReqT>() {
+          @Override
+          public void onMessage(ReqT message) {
+            delegate.onMessage(message);
+          }
+
+          @Override
+          public void onHalfClose() {
+            delegate.onHalfClose();
```

Comment:
I'm hazy on these callbacks: does a half close precede a onCancel/onComplete? Do you need to decrement the counter here?

Reply:
I was a bit confused too TBH so I checked the docs:
```<p>Any contexts are guaranteed to arrive before any messages, which are guaranteed before half close, which is guaranteed before completion.``` (https://github.com/grpc/grpc-java/blob/80c3c992a66aa21ccf3e12e38000316e45f97e64/api/src/main/java/io/grpc/ServerCall.java#L45)